cmd/private-prow-configs-mirror/main.go

```diff
@@ -0,0 +1,437 @@
+			newValue := value
+
+			// We want to append the org level values, because the private repo
+			// can exist in different org with different configuration.
+			org := strings.Split(orgRepo, "/")[0]
+			if _, ok := plugins[org]; ok {
+				newValue = append(newValue, plugins[org]...)
+			}
+
+			privateReposWithValues[privateOrgRepo(strings.Split(orgRepo, "/")[1])] = newValue
+			allPrivate = append(allPrivate, newValue...)
+		}
+	}
+
+	mergeCommonPrivatePlugins(plugins, privateReposWithValues, getRepeatedValues(allPrivate))
+}
+
+// mergeCommonPrivatePlugins detects the common values from all the private repositories
+// and merges them into the org `openshift-priv` level.
+// In addition, it generates a repo. if it contains an extra non-common field.
+func mergeCommonPrivatePlugins(plugins map[string][]string, privateReposWithValues map[string][]string, repeatedValues sets.String) {
+	privateOrgPlugins := sets.NewString()
+
+	for repoName, repoValues := range privateReposWithValues {
+		repoLevelPlugins := sets.NewString()
+		valuesSet := sets.NewString(repoValues...)
+
+		if len(repeatedValues) > 0 {
+			commonValues := valuesSet.Intersection(repeatedValues)
+			privateOrgPlugins.Insert(commonValues.List()...)
+
+			if diffValues := valuesSet.Difference(repeatedValues); len(diffValues) > 0 {
+				repoLevelPlugins.Insert(diffValues.List()...)
+			}
+
+		} else {
+			repoLevelPlugins.Insert(repoValues...)
+		}
+
+		if len(repoLevelPlugins.List()) > 0 {
+			logrus.WithFields(logrus.Fields{"repo": repoName, "value": repoLevelPlugins.List()}).Info("Generating repo")
+			plugins[repoName] = repoLevelPlugins.List()
+		}
+	}
+
+	if len(privateOrgPlugins.List()) > 0 {
+		logrus.WithField("value", privateOrgPlugins.List()).Info("Generating openshift-priv org.")
+		plugins[openshiftPrivOrg] = privateOrgPlugins.List()
+	}
+}
+
+func getRepeatedValues(values []string) sets.String {
+	temp := sets.NewString()
+	repeated := sets.NewString()
+
+	for _, value := range values {
+		if temp.Has(value) {
+			repeated.Insert(value)
+		} else {
+			temp.Insert(value)
+		}
+	}
+
+	return repeated
+}
+
+func main() {
+	logrus.SetReportCaller(true)
+	logrus.SetFormatter(&logrus.TextFormatter{
+		CallerPrettyfier: func(f *runtime.Frame) (string, string) {
+			return fmt.Sprintf("%s()", f.Function), fmt.Sprintf("%s:%d", path.Base(f.File), f.Line)
+		},
+	})
+
+	o := gatherOptions()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid option")
+	}
+
+	go func() {
+		interrupts.WaitForGracefulShutdown()
+		os.Exit(1)
+	}()
+
+	currentDir, err := os.Getwd()
+	if err != nil {
+		logrus.WithError(err).Fatal("couldn't get current directory")
+	}
+
+	configs := config.GetAllConfigs(currentDir, logrus.NewEntry(logrus.New()))
+	pluginsConfig, err := loadProwPlugins(o.prowPlugins)
+	if err != nil {
+		logrus.WithError(err).Fatal("could not load Prow plugin configuration")
+	}
+	prowConfig := configs.Prow.ProwConfig
```

Comment:
This will NPD if that config failed to be read, see the doc for `GetAllConfigs`.